User Lens Calibration Screen + Lens / Corner decoupling

amrsoll opened the pull request:
## Expected behaviour for the User Lens Calibration:
The "Start" button in the "Lens Calibration" widget sets the device in lens calibration mode.
- The button is used to take pictures
- No marker detection is done
- Work space image is Not updated
- Exit calibration if the user leaves lens calibration page or presses one of the cancel buttons
- Delete unused images for the lens calibration when cancelling.
- Deletes images when the user clicks on factory reset.

#### Lights

- Solid :blue_square:  Idle, waiting to take a picture
- Flashing  (fast) :blue_square: Taking a picture
- Breathing (slow) :blue_square: Minimum pictures needed :heavy_check_mark:  + detecting boards + can take new picture
- Flashing :green_square: Taking picture, Obtained minimum number pics **or** Calculating the lens distortion based on detected boards
- Solid :green_square: Lens Calibration complete ; waiting to take further pictures.

#### Abandoned changes

- Check result during the calibration > Take new pic + undistort & refresh work area
  - Take a new picture each time the calibration completes
  - Refresh button in the work area
  - Allows the user to compare the result between calibrations
    1. Laser patterns before starting lens calibration
    2. Start lens calibration
    3. Between successful calibrations, go back compare your image result to the expected design
    4. Stop when satisfactory

## Internal changes

- Most position values have been converted from `(i, j)` to `(x, y)` (inverted, OpenCV standard)
- Lens and Corner calibration are independent
- Can track a point after a lens distortion
- Performance : 
  - Marker detection on Raw image
  - No marker detected on raw img + precision mode -> No lens undistortion

#### Pushed back for after merging this :

- Make the undistortion resolution-independent  (Too buggy to make work yet.)
  - Work-around : Detect markers on raw high-res pic -> downscale -> undistort (also saves cpu time) 
  - Only interesting if we decide to scale up img res (no need yet - precision isn't there)
  - Necessary if we do the lens calibration with higher res pictures. -> Better calibration
- Notify frontend when the backend is ready to start lens calibration again. (semi-complete)
- Calibration Tool mode
  - Change the order of corner and lens calibration

Fixes #867

irlaec committed:
Remove button to take pictures

irlaec committed:
wip

amrsoll committed:
Decoupling : Stage 1

- Detect markers on the raw picture
- Undistort the picture after the detection
- Add some dictionnary tools
- Add documentation
- Can now predict a pixel position after undistort
- Possibility to up/downscale image
- TODO : Could take a different picture resolution.

TODO : Calibraton tool should save:
- the raw position of the arrow tip on the image
- Convert that position through lens distortion for backwards compatibility

amrsoll committed:
Decoupling: Stage 2

- Corner calibration saves in the correct locations
- Optimisation : Turn the image grey before lens correction
- yaml dump should use the default (more readable) dumper

irlaec committed:
Lens calibration page style

amrsoll committed:
Decoupling : Stage 3

- Refactoring
  - definitions file to store constants separately
  - corner and lens calibration separated into 2 files
  - lid handler : serve picture gets shorter
- Clean up some file saving feature

amrsoll committed:
Decoupling : Final stage

- All the constants imported from the definitions file
- MrbPicWorker defined in worker.py
- Removed typing dependency (unavailable on our machines)
- Moved lens undistorion functions from init and undistort.py to lens.py

amrsoll committed:
Change coords to x,y instead of i,j + dbg imgs.

amrsoll committed:
Add camera unit test with pytest.

amrsoll committed:
Move camera worker to separate module + pytest.ini

amrsoll committed:
Fix i j to x y coords.

amrsoll committed:
Rm test outputfile

amrsoll committed:
Ignore debug files from tests.

amrsoll committed:
Extra debug images and pytest debug.

amrsoll committed:
Merge remote-tracking branch 'origin/f_user_lens_calibration' into f_combined_user_lens_calibration

amrsoll committed:
Check bindings & small refactors + missing picture sort.

amrsoll committed:
Disable arrow lasering when not homed.

amrsoll committed:
Lens calibration fixes + test functions.

irlaec committed:
wip: viewmodel refactoring

amrsoll committed:
Add optional starting state + auto rm board pics for PROD.

amrsoll committed:
+ Assure cleaning of board pictures.

amrsoll committed:
Prevent slicing when calibrating the lens.

amrsoll committed:
verbosity++

amrsoll committed:
update pytest.ini

amrsoll committed:
Fix bug where calibration state wouldn't refresh.

irlaec committed:
Split calibration/camera settings in different view models

irlaec committed:
Adapt for Watterott calibration screen

irlaec committed:
Revert watterott mode by default (it was for development)

amrsoll committed:
Merge branch 'f_combined_user_lens_calibration' into f_lens_calib_viewmodels

amrsoll committed:
Fix binding, comments and indentation.

irlaec committed:
Merge pull request #974 from mrbeam/f_lens_calib_viewmodels

Split calibration in different view models

amrsoll committed:
convert i, j -> x y by default.

amrsoll committed:
Explicitely abort the lens calibration.

- Do not abort using the "Back" button.
- Do not abort when the settings menu is shown.

amrsoll committed:
Factory vs user lens calibration file.

amrsoll committed:
Lens calibration perf + modify testing

- During len calibration, Camera waits for the user pressing on the button.

ghost commented:
## DeepCode's analysis on [#594cc4](https://www.deepcode.ai/app/gh/mrbeam/MrBeamPlugin/4286ee5fd893d1b9d468d0c651bdf3bd085a0a97/mrbeam/MrBeamPlugin/594cc49bd517e02758face14bcb7698aceaf4098/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=2&) found:

- :information_source: **2** minor issues. :point_down:
- :heavy_check_mark: **4** issues were fixed.
## Top issues

<table>
  <thead>
    <tr>
      <th align="left">Description</th>
      <th align="left">Example fixes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td width="77%">No need to call keys, directly check with the in operator. Occurrences: <ul><li><a href="https://github.com/mrbeam/MrBeamPlugin/blob/594cc49bd517e02758face14bcb7698aceaf4098/octoprint_mrbeam/camera/corners.py#L144">corners.py:144</a></li> <li><a href="https://github.com/mrbeam/MrBeamPlugin/blob/594cc49bd517e02758face14bcb7698aceaf4098/octoprint_mrbeam/camera/lens.py#L358">lens.py:358</a></li> <li><a href="https://github.com/mrbeam/MrBeamPlugin/blob/594cc49bd517e02758face14bcb7698aceaf4098/octoprint_mrbeam/camera/lens.py#L512">lens.py:512</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/mrbeam/MrBeamPlugin/4286ee5fd893d1b9d468d0c651bdf3bd085a0a97/mrbeam/MrBeamPlugin/594cc49bd517e02758face14bcb7698aceaf4098/pr/_/%2Foctoprint_mrbeam%2Fcamera%2Fcorners.py/python%2Fdc%2Freplace~keys~list~compare/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
    <tr>
      <td width="77%">Using the loop variable _k from some outer or global scope in this for of loop may prevent optimisations by the JavaScript engine. Consider declaring the loop variable inside the head of this for of loop. Occurrences: <ul><li><a href="https://github.com/mrbeam/MrBeamPlugin/blob/594cc49bd517e02758face14bcb7698aceaf4098/octoprint_mrbeam/static/js/calibration/corner_calibration.js#L223">corner_calibration.js:223</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/mrbeam/MrBeamPlugin/4286ee5fd893d1b9d468d0c651bdf3bd085a0a97/mrbeam/MrBeamPlugin/594cc49bd517e02758face14bcb7698aceaf4098/pr/_/%2Foctoprint_mrbeam%2Fstatic%2Fjs%2Fcalibration%2Fcorner_calibration.js/javascript%2Fdc%2FNonLocalLoopVar/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
  </tbody>
</table>


#### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/mrbeam/MrBeamPlugin/4286ee5fd893d1b9d468d0c651bdf3bd085a0a97/mrbeam/MrBeamPlugin/594cc49bd517e02758face14bcb7698aceaf4098/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=2&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=mrbeam)

amrsoll committed:
Lens calibration pretty stable.

- more than 9 pic for users
- cancel the calibration is slightly more stable
- calibration waits for you to press on the btn to take picture
- Front tend message (starting / finished calibration)
- No marker detection during lens calibration
- *Need* to press on the "cancel" btn to cancel the calibration
- Zoom icons don't show on picture placeholders

amrsoll committed:
Merge remote-tracking branch 'origin/develop' into f_combined_user_lens_calibration

amrsoll committed:
Fix unreferenced variable.

amrsoll committed:
fixes.

amrsoll committed:
Switch solid blue when no calibration is IDLE.

amrsoll committed:
Back button also cancels lens calibration.

amrsoll committed:
Revert lens calibration -> factory.

amrsoll committed:
Fix factory revert.

irlaec committed:
Abort calibrations when going to another settings tab

amrsoll committed:
Lens calibration FINALLY exits immediatly.

Turns out process.terminate() is a SIGTERM and not SIGKILL ...

UI : First attempt at greying out the lens calibration start button
while the backed isn't ready.

amrsoll committed:
Save & rm factory pictures + bugfix.

amrsoll committed:
minor sort fix.

amrsoll committed:
Simplify marker saveFile and add version string.

irlaec committed:
Revert back button styling

irlaec committed:
Remove todos

irlaec committed:
Bring back calibration tool overlay

amrsoll committed:
Remove unused title class.
